File: DataStructures/BinaryTree/maxDepth.js
/*
* Definition for a binary tree node.
* function TreeNode(val, left, right) {
*     this.val = (val===undefined ? 0 : val)
*     this.left = (left===undefined ? null : left)
*     this.right = (right===undefined ? null : right)
* }
*/
/**
* @param {TreeNode} root
* @return {number}
*/

// Top -> Bottom (Preorder Traversal)
var maxDepthTopToBottom = function(root) {
   if(root === null) return 0;
   let answer = 0;
   const preorderTraversal = (root, depth = 0) => {
       if(root === null) return;
       
       answer = Math.max(answer, depth + 1);
       if(root.left !== null){
           preorderTraversal(root.left, depth + 1);
       }
       if(root.right !== null){
           preorderTraversal(root.right, depth + 1);
       }
       
   }
   
   preorderTraversal(root);
   
   return answer;
};

File: Leetcode/BinaryTrees/maxDepth.py
# 104. Maximum Depth of Binary Tree
# https://leetcode.com/problems/maximum-depth-of-binary-tree

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
# Bottom -> Up
class Solution:
    def maxDepth(self, root: TreeNode, depth=0) -> int:
        if(root == None):
            return depth

        left = self.maxDepth(root.left, depth + 1)
        right = self.maxDepth(root.right, depth + 1)

        return max(left, right)
